### PR TITLE
Feature/unit test

### PR DESCRIPTION
**Pull request by mersanuzun**

### CODE REVIEW

**Comment by fatihacet** [comment]
@mersanuzun that's a great contribution, thanks for doing this 👍 I was thinking to do it from the beginning but there was really no time before WIP.

@abdullah can you review this one for me?

**Comment by mersanuzun** [comment]
@fatihacet  thank you. :) It is a really good repository. I just started with 2 components but it is just a beginning :)

File: package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
     "build": "vue-cli-service build",
     "lint": "vue-cli-service lint",
     "test:unit": "vue-cli-service test:unit",
+    "test:clear:snapshot": "npm run test:unit -- -i -u",
```
Comment (abdullah) [inline package.json]: When I run this command with `yarn`, I get an error `.... Use the `--scripts-prepend-node-path` option to include the path for the node binary npm was executed with.`. Can you resolve it?

File: jest.config.js [hunk 2]
```diff
@@ -8,8 +8,11 @@ module.exports = {
   transform: {
     '^.+\\.vue$': 'vue-jest',
     '.+\\.(css|styl|less|sass|scss|svg|png|jpg|ttf|woff|woff2)$': 'jest-transform-stub',
-    '^.+\\.jsx?$': 'babel-jest',
+    '^.+\\.(js|jsx)?$': '<rootDir>/node_modules/babel-jest',
```
Comment (abdullah) [inline jest.config.js]: Why did you change this line ? I tried old regex and path and it works.

File: tests/unit/src/components/shared/AppBanner.spec.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
+import { shallowMount, createLocalVue } from '@vue/test-utils';
```
Comment (abdullah) [inline tests/unit/src/components/shared/AppBanner.spec.js]: `createLocalVue` is not used in this file.

File: tests/unit/src/components/shared/AppFooter.spec.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
+import { shallowMount, createLocalVue } from '@vue/test-utils';
```
Comment (abdullah) [inline tests/unit/src/components/shared/AppFooter.spec.js]: Same problem here 😄

File: tests/unit/src/components/shared/AppFooter.spec.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+import AppFooter from '@/components/shared/AppFooter.vue';
+
+describe('AppFooter.vue', () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  beforeAll(() => {
+    Object.defineProperty(window, 'localStorage', {
```
Comment (abdullah) [inline tests/unit/src/components/shared/AppFooter.spec.js]: It can be moved to setup file. WDYT @mersanuzun ?

File: tests/unit/src/components/shared/AppFooter.spec.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  });
+
+  describe('data', () => {
+    it('should have "isBackToTopVisible" as false value', () => {
+      const componentState = AppFooter.data();
+      
+      expect(componentState.isBackToTopVisible).toEqual(false);
+    });
+
+    it('should have "isFooterSubscribeVisible" as true value if "subscribeClosed" value is falsey', () => {
```
Comment (abdullah) [inline tests/unit/src/components/shared/AppFooter.spec.js]: Spelling mistake. End of the sentence should be `false` not `falsey`.

File: tests/unit/src/components/shared/AppFooter.spec.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+      const componentState = AppFooter.data();
+      
+      expect(componentState.isFooterSubscribeVisible).toEqual(true);
+    });
+
+    it('should have "isFooterSubscribeVisible" as true value if "subscribeClosed" value is falsey', () => {
```
Comment (abdullah) [inline tests/unit/src/components/shared/AppFooter.spec.js]: Same spelling mistake. 😅

File: tests/unit/src/components/shared/AppFooter.spec.js [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+      });
+    }
+
+    it('should match the snapshot', () => {
+      const wrapper = shallowAppFooter();
+
+      expect(wrapper.vm.$el).toMatchSnapshot();
+    });
+
+    it('should not render .has-widget div which its id is #footer, if subscribeClosed is true', () => {
```
Comment (abdullah) [inline tests/unit/src/components/shared/AppFooter.spec.js]: I prefer `js` prefix while using DOM elements on js side. It would be better. WDYT @fatihacet @mersanuzun  ?

File: tests/unit/src/components/shared/AppFooter.spec.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+        wrapper.vm.handleScroll();
+
+        expect(wrapper.vm.isBackToTopVisible).toEqual(false);
+      });
+    });
+  });
+
+  describe('templates', () => {
+    const shallowAppFooter = () => {
```
Comment (abdullah) [inline tests/unit/src/components/shared/AppFooter.spec.js]: Why didn't you use `beforeEach` hook? Also, each component should be destroyed after each `describe` scope to prevent memory leak. You can use `afterEach` hook for it.

**Comment by abdullah** [comment]
@mersanuzun I've completed initial review. Good work. 💪 Thank you 👍
In addition, there are some eslint problems. You can see these errors via run `yarn/npm run lint`. 

I did not review components' logic. I will review ASAP.

**Comment by abdullah** [comment]
IMHO, the commit messages should be imperative tense and start with capital letter. Like this https://github.com/kodilan-com/frontend/commit/5ddb70fc479537ba60728ae0575eef7d570b8bf1.

File: jest.config.js (hunk 2 above)
Comment (mersanuzun) [reply jest.config.js]: When I was writing tests, It did not work on my environment. I changed pc and install all dependencies again, now It is working without `<rootDir>/node_modules`. :)

File: tests/unit/src/components/shared/AppFooter.spec.js (hunk 8 above)
Comment (mersanuzun) [reply tests/unit/src/components/shared/AppFooter.spec.js]: I don't understand the issue :)

Comment (abdullah) [reply tests/unit/src/components/shared/AppFooter.spec.js]: There is no issue :) It is a recommendation.

**Comment by mersanuzun** [comment]
@abdullah I created `generateWrapper` function for the assigning mock issue. But you are right we can create a `localWrapper` function for this case.

File: tests/unit/src/components/shared/AppFooter.spec.js [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+      const componentState = AppFooter.data();
+
+      expect(componentState.isFooterSubscribeVisible).toEqual(false);
+    });
+  });
+
+  describe('methods', () => {
+    describe('handleScroll', () => {
+      const setScrollY = (value) => {
```
[the remark this answers is not in the document]
Comment (mersanuzun) [reply tests/unit/src/components/shared/AppFooter.spec.js]: Done :) @abdullah

**Comment by abdullah** [comment]
@mersanuzun Thanks for your effort and contributions. 💪 

I have some question for the Test setup and CI. 

@mersanuzun Please comment your opinion. 👍 
@fatihacet Can you answer too?

- [ ] Will we report the coverage? We will just put ` --coverage`  end of the test script 😅 
- [ ] Should we use `js` prefix while using DOM element on `JavaScript` side? It would be better while reading the test code. 

For instance;  
`wrapper.find('input.ico-01') => wrapper.find('.js-search-input')` 

Here is @mersanuzun coverage 💯 

```
-----------------------|----------|----------|----------|----------|-------------------|
File                   |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
-----------------------|----------|----------|----------|----------|-------------------|
All files              |    53.85 |    13.33 |      100 |    53.85 |                   |
 src/assets/data       |      100 |      100 |      100 |      100 |                   |
  locations.js         |      100 |      100 |      100 |      100 |                   |
 src/components/shared |    70.83 |       50 |      100 |    70.83 |                   |
  AppBanner.vue        |      100 |      100 |      100 |      100 |                   |
  AppFooter.vue        |      100 |      100 |      100 |      100 |                   |
  AppHeader.vue        |      100 |      100 |      100 |      100 |                   |
  AppTopbar.vue        |        0 |      100 |      100 |        0 |           4,10,11 |
  ApplyLink.vue        |      100 |      100 |      100 |      100 |                   |
  LocationSelect.vue   |    33.33 |        0 |      100 |    33.33 |       19,26,32,36 |
 src/utils             |        0 |        0 |      100 |        0 |                   |
  query.js             |        0 |        0 |      100 |        0 |... 14,15,18,19,22 |
 tests/unit/src        |      100 |      100 |      100 |      100 |                   |
  setupTests.js        |      100 |      100 |      100 |      100 |                   |
-----------------------|----------|----------|----------|----------|-------------------|
```

**Comment by fatihacet** [comment]
@mersanuzun AMAZING WORK 💪 Thanks for working on this one 👍 It's awesome to see 53.85% coverage. 

@abdullah thank you for reviewing this. I can do Travis integration and we will definitely have coverage reports. Also, we can add `js-` prefix later. It's not a big deal.

I will go ahead and merge this one. Thanks everyone 👍
